mart/attack/enforcer.py

```diff
@@ -97,17 +95,43 @@ def verify(self, input_adv, *, input, target):
 
 
 class Enforcer:
-    def __init__(self, constraints: dict[str, Constraint] | None = None) -> None:
-        self.constraints = constraints or {}
+    def __init__(self, **modality_constraints: dict[str, dict[str, Constraint]]) -> None:
+        self.modality_constraints = modality_constraints
 
     @torch.no_grad()
+    def _enforce(
+        self,
+        input_adv: torch.Tensor,
+        *,
+        input: torch.Tensor,
+        target: torch.Tensor | dict[str, Any],
+        modality: str,
+    ):
+        for constraint in self.modality_constraints[modality].values():
+            constraint(input_adv, input=input, target=target)
+
     def __call__(
         self,
-        input_adv: torch.Tensor | tuple,
+        input_adv: torch.Tensor | tuple | list[torch.Tensor] | dict[str, torch.Tensor],
         *,
-        input: torch.Tensor | tuple,
-        target: torch.Tensor | dict[str, Any] | tuple,
+        input: torch.Tensor | tuple | list[torch.Tensor] | dict[str, torch.Tensor],
+        target: torch.Tensor | dict[str, Any],
+        modality: str = "constraints",
         **kwargs,
-    ) -> None:
-        for constraint in self.constraints.values():
-            constraint(input_adv, input=input, target=target)
+    ):
+        assert type(input_adv) == type(input)
+
+        if isinstance(input_adv, torch.Tensor):
+            # Finally we can verify constraints on tensor, per its modality.
+            # Set modality="constraints" by default, so that it is backward compatible with existing configs without modalities.
+            self._enforce(input_adv, input=input, target=target, modality=modality)
+        elif isinstance(input_adv, dict):
+            # The dict input has modalities specified in keys, passing them recursively.
+            for modality in input_adv:
+                self(input_adv[modality], input=input[modality], target=target, modality=modality)
+        elif isinstance(input_adv, (list, tuple)):
+            # The list or tuple input is a collection of sub-input and sub-target.
+            for input_adv_i, input_i, target_i in zip(input_adv, input, target):
+                self(input_adv_i, input=input_i, target=target_i)
```

Comment:
You might want to `assert modality == "constraints"` here.

Reply:
Thanks. The assertion is added.

I also added `modality=modality` to the recursive call so that we pass along the modality.

We can simply remove the assertion later if we want to loosen the data structure constraints (e.g. allow list/tuple in a modality-dictionary).